### DISCUSSION TITLE
Combining continuous and dichotom variables

### DISCUSSION THREAD

**Discussion by max-kat** [rsagroup/rsatoolbox#409] [Q&A]
Dear community, 

I am stuck on a problem for quite some time now and would really appreciate your thoughts on this:

My experiment has a 2x2 design. Participants have to **imagine**  stimuli (condition A) and are **shown** stimuli (condition B). These stimuli consist of 2 types: **Faces and Artworks**. 
After each presentation participants judged the evoked experience on how 1. pleasurable, 2. beautiful, and 3 moving it was. 
**Furthermore, only in the imagery condition they also evaluate how vivid the imagined picture was.** 

My research question is whether the neuronal representations underlying the elicited aesthetic experiences (pleasure/beauty/moving) across the different conditions (imagery vs perception) and stimulus types (faces vs artworks) are similar (or not). 

In the first step, I created the single trial beta maps for each condition. In the next step I created a candidate model for perception vs. imagery and one for face vs artwork. These are dichotomous and easy to generate. 
But then it gets difficult. If I just focus on one experience (e.g. moving) the candidate model is no longer dichotomous but now continuous (between 1-7). However, I guess I just use the raw values and then create a dissimilarity matrix. However, I am not sure how I combine all 3 ratings into one RDM that I could test. Could someone explain to me how this is done? 
Furthermore, I have the suspicion that elicited experiences become more similar between perception and imagery when vividness was high and get more dissimilar when they are not very vivid. However, since vividness was only assessed for imagery, I have no clue how to incorporate this assumption into an RDM cause I only have values for imagery and none for perception. 
I would really appreciate any help on this! 
Also if you have time and really want to work with me on this, I'd be open for a collaboration! 

Thanks for reading this far!

All the best, 
Max

**Comment by JasperVanDenBosch** [comment]
Hi Max,

> Furthermore, I have the suspicion that elicited experiences become more similar between perception and imagery when vividness was high and get more dissimilar when they are not very vivid. However, since vividness was only assessed for imagery, I have no clue how to incorporate this assumption into an RDM cause I only have values for imagery and none for perception.
> I would really appreciate any help on this!

For this maybe you could split the data for imagery into several RDMs according to vividness, e.g. a low-, mid and high- vividness imagery RDM. (Take the low-vididness trials and create an RDM from that, etc) You could then compare the perception RDM with these one by one?

> However, I am not sure how I combine all 3 ratings into one RDM that I could test. Could someone explain to me how this is done?

Here I think it depends on exactly what you want to hypothesise. Have you considered a [MixedModel](https://rsatoolbox.readthedocs.io/en/stable/model_specification.html#weighted-models)? You could either set static weights, e.g, `task * 0.33 + moving * 0.33 + face vs art 0.33`, or even have rsatoolbox determine flexible weights when fitting the model.

Hope this helps!

**Reply by max-kat** [reply]
Thank you so much for responding @JasperVanDenBosch! 

I thought about it for 2 days and decided to create single contrast estimates for the 8 conditions for each participant only using trials with at least 6+7 in vividness. 

However, this results in a bit of a problem:
not all participants have all 8 conditions. This means that some subject-level rdms are only e.g. 5x5 instead of 8x8.
I then used partial_rdms = from_partials(listing_individual_rdms) to cast them into the same shape with nan if the condition was missing. 
plot of the individual rdms
![image](https://github.com/user-attachments/assets/22b27d59-b700-402c-b934-31ada4d69c68)

However, I do not find a function on how to now infer which of my fixed_models fits the data best cause it seems that the nan are breaking the evaluation... 

neuro_results_1 = rsatoolbox.inference.evaluate.eval_fixed(models, partial_rdms, method='cosine_cov')
The above command does not work

How can I fix this? Or is there no way to infer which model is best when you have nan in your data?

**Reply by max-kat** [reply]
P.S. Is the only way to create a weighted mean RDM of all the partial ones and then do:
rescaledPartials = rescale(partial_rdms, method='evidence')
meanRDM = rescaledPartials.mean(weights='rescalingWeights') ?

neuro_results_1 = rsatoolbox.inference.evaluate.eval_fixed(models, meanRDM) 
rsatoolbox.vis.plot_model_comparison(neuro_results_1 ) --> but this has no variance and can thus not be plotted. also because it has no variance how could I say one is better then the other since no variance also means no p-value in model comparison?

**Reply by JasperVanDenBosch** [reply]
That's right, the evaluation functions currently require nan-free rdms.
Indeed the solution would be either to test on the subsets, or to combine them somehow.
One way would be to apply a regular `nanmean`; but  this assumes that the scale and distribution of the partial RDMs is the same. The `rescale` function is for those cases where you cannot make that assumption. One example is the Multiple Arrangement task where every trial tests a subset of stimuli, such that the participant's context is different every trial.
In order to end up with multiple data RDMs, is there a way for you to subset the data such that you can create fully defined RDMs? E.g. groups of 3 participants?

**Reply by max-kat** [reply]
thanks for the explanation!

regarding

>  In order to end up with multiple data RDMs, is there a way for you to subset the data such that you can create fully defined RDMs? E.g. groups of 3 participants?

the idea is to mix and match individual participants to form one complete rdm right? So if one participant misses rdm a and b but has c and d and another person has a&b but misses c&d, I could create a mean rdm for their combined data. Since I do this for subgroups of 2/3 participants I end up with e.g. 15-20 combined RDMs which I then could evaluate with fixed_model --> which now would give me variance estimates and p-values for model comparisons as I entered >1 RDMs for the calculation, right?
Did I understand your approach correctly? 
(btw thanks for taking the time!!!)

**Reply by JasperVanDenBosch** [reply]
That's the general idea yes, but I wouldn't take it that far. If you can't make a complete (abcd) RDM from say, a sub-group of 1/5 of your participants, I don't think this approach is the best one to take. In that case you may be best of a) testing hypotheses on the smaller subset-rdms or b) just reporting the group RDM results without stats.

**Reply by max-kat** [reply]
Thanks now it is clearer! However, I think I will for now stick with the very restrictive subgroup of full cases for each participant which results in 11 individual rdms. Interestingly, they almost exactly mirror the overall data analysis without taking vividness into account (except for some very minor reduction in explained accuracy by the modality model (perception vs imagery). 
mhm so if I would rerun the analysis for low vividness and coincidentally the same 11 participants would pop up with full cases in the analysis. is there a way to quantitatively say something about the difference vividness causes between rdms? or would I only be able to qualitatively explain that e.g. the accuracy of model A is in the high vividness analysis higher than in the low vividness group?

**Reply by JasperVanDenBosch** [reply]
I'd say reporting the difference in model fit values is quantitative :)

**Reply by max-kat** [reply]
Yes sorry I was unsure whether one needs to run an extra test afterwards if let's say a difference between 8% in model performance between the low vs high vividness data set is 'significant' or not! but if not then it's fine! 
Thanks for everything!!

**Reply by JasperVanDenBosch** [reply]
Btw I know someone I think you should talk to. Would you send me your email address somewhere? I'll then introduce you if you haven't talked yet.

**Reply by max-kat** [reply]
maximilian.kathofer@univie.ac.at thanks for connecting me!!!

**Reply by max-kat** [reply]
one last question regarding your previous answer:

> If you can't make a complete (abcd) RDM from say, a sub-group of 1/5 of your participants, I don't think this approach is the best one to take. 

Do you mean that if I would need more than 1/5 of participants to form a full RDM, this approach wouldn't be great? 
So, if I just use groups of 2 participants (2/33), it would be ok? I would just match them based on what they miss & have, so I would need only 2 participants on average to form a complete rdm. since some have complete cases as well I would roughly double my 'available' data. would that be a somewhat sound approach?

**Reply by JasperVanDenBosch** [reply]
Exactly. With 2/33 you'd end up with 16 RDMs which is reasonable to get some measure of variance. If you only had say, 4 RDMs that would kind of defeat the purpose. Note that you'd not be able to generalise across the population, since these RDM's do not stand for individuals but rather mini-groups. But at least you'd have some sense of the variance. You could also resample the participant combinations.

**Reply by max-kat** [reply]
Sorry, @JasperVanDenBosch one last conceptual question: 

If I have 2 runs and a 2x2 design and all conditions are present across both runs, I can use the crossnobis distance to estimate similarities. 
However, what if participants also rated the stimuli afterward (e.g. high & low for simplicity)? 
Now let's say all high trials were in run 1 and all low trials were in run 2. Is it still 'valid' to use crossnobis distance? Cause for me - at least theoretically- the design changed from 2x2 across 2 runs to a 2x2x2 design (the last 2 are high and low) and thus crossnobis can no longer be used since high and low are not present across both runs. More specifically, I suspect that crossnobis might consider the evoked experience across runs now as noise (as the experiences are not present across runs to validate) and thus similarity estimates for the evoked experience would be attenuated. Is this a valid fear? Or can I still use crossnobis since the design initially was a 2x2 design?

**Reply by JasperVanDenBosch** [reply]
In this case you wouldn't have runs on which to crossvalidate so crossnobis wouldn't make sense (it would be mahalanobis).

PS Can you create new topics for subsequent questions? That makes it easier for other users to find their answers :)
